craft_parts/plugins/java_plugin.py

```diff
@@ -0,0 +1,122 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2020-2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Plugin base class and definitions."""
+
+from __future__ import annotations
+
+import logging
+import os
+import subprocess
+import tempfile
+
+from overrides import override
+
+from .base import Plugin
+
+logger = logging.getLogger(__name__)
+
+
+class JavaPlugin(Plugin):
+    """A base class for java-related plugins.
+
+    Provide common methods to deal with the java executable location and
+    symlink creation.
+    """
+
+    def _check_java(self, javac: str) -> tuple[int | None, str]:
+        with tempfile.TemporaryDirectory() as tempdir:
+            test_class = """
+                public class Test {
+                    public static void main(String[] args){
+                        System.out.println(System.getProperty("java.specification.version"));
+                    }
+                }"""
+            with open(f"{tempdir}/Test.java", "w") as file:
+                file.write(test_class)
+
+            try:
+                subprocess.call([javac, "-d", tempdir, f"{tempdir}/Test.java"])
+                java_home = os.path.dirname(os.path.dirname(javac))
+                spec_version = subprocess.check_output(
+                    [java_home + "/bin/java", "-cp", tempdir, "Test"]
+                )
+                # Java 8 reports spec 1.8. Treat it as a spec version 8
+                # 11 and up report the actual spec version number
+                version = int(spec_version.decode().split(".")[-1])
+
+            except subprocess.CalledProcessError as err:
+                logging.info(f"{javac} is not a valid Java compiler: {err.output}")
+            except PermissionError as err:
+                logging.info(
+                    f"{javac} is not a valid Java compiler. Permission error {err}"
+                )
+            else:
+                return version, java_home
+        return None, ""
+
+    def _find_javac(self) -> list[str]:
+        cmd = ["find", "/usr/lib/jvm", "-path", "*/bin/javac", "-print"]
+        output = subprocess.check_output(cmd)
```

Comment:
```suggestion
        output = subprocess.check_output(cmd, text=True)
```

Then you don't need to decode()